apollo-relay/src/lib.rs

```diff
@@ -0,0 +1,532 @@
+                let client = Client::new();
+                let mut interval = interval(Duration::from_secs(5));
+                interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
+                interval.tick().await;
+                loop {
+                    tokio::select! {
+                        biased;
+                        mopt = rx.recv() => {
+                            match mopt {
+                                Some(_msg) => {
+                                    relay_tpq(&client, task_tpq.clone()).await;
+                                },
+                                None => break
+                            }
+                        },
+                        _ = interval.tick() => {
+                            relay_tpq(&client, task_tpq.clone()).await;
+                        }
+                    };
+                }
+            });
+            Self { addr, tpq, tx }
+        }
+
+        /// Start serving requests.
+        pub async fn serve(self) -> Result<(), Error> {
+            let addr = self.addr;
+            Server::builder()
+                .add_service(ReporterServer::new(self))
+                .serve(addr)
+                .await
+        }
+
+        async fn submit_report(
+            client: &Client,
+            key: String,
+            report: report::Report,
+        ) -> Result<Response<ReporterResponse>, Status> {
+            tracing::debug!("submitting report: {:?}", report);
+            // Protobuf encode message
+            let mut content = BytesMut::new();
+            report
+                .encode(&mut content)
+                .map_err(|e| Status::invalid_argument(e.to_string()))?;
+            // Create a gzip encoder
+            let mut encoder = GzEncoder::new(Vec::new(), Compression::default());
+            // Write our content to our encoder
+            encoder
+                .write_all(&content)
+                .map_err(|e| Status::internal(e.to_string()))?;
+            // Finish encoding and retrieve content
+            let compressed_content = encoder
+                .finish()
+                .map_err(|e| Status::internal(e.to_string()))?;
+            let mut backoff = 0;
+            for i in 0..4 {
+                let res = client
+                    .post("https://usage-reporting.api.apollographql.com/api/ingress/traces")
+                    .body(compressed_content.clone())
+                    .header("X-Api-Key", key.clone())
+                    .header("Content-Encoding", "gzip")
+                    .header("Content-Type", "application/protobuf")
+                    .header("Accept", "application/json")
+                    .send()
+                    .await;
+                match res {
+                    Ok(_v) => break,
+                    Err(e) => {
+                        tracing::warn!("attempt: {}, could not trigger transfer: {}", i + 1, e);
+                        backoff += 500;
```

Comment:
The TS implementation retries on 5xx but not on 4xx, since the latter typically means there's something wrong with the message itself.

Reply:
Ok. I can clean this up as required.